File: src/main/java/io/disassemble/javanalysis/flow/ControlFlowGraph.java
package io.disassemble.javanalysis.flow;

import javassist.bytecode.analysis.ControlFlow;

import java.util.*;

/**
 * @author Tyler Sedlar
 * @since 5/20/2017
 */
public class ControlFlowGraph {

    private final ControlFlow flow;
    private final List<FlowBlock> blocks;

    public ControlFlowGraph(ControlFlow flow, List<FlowBlock> blocks) {
        this.flow = flow;
        this.blocks = blocks;
    }

    public ControlFlow flow() {
        return flow;
    }

    public List<FlowBlock> blocks() {
        return blocks;
    }

    public Collection<FlowNode> dominatorTree() {
        Map<ControlFlow.Node, FlowNode> nodes = new LinkedHashMap<>();
        for (ControlFlow.Node node : flow.dominatorTree()) {
            nodes.put(node, new FlowNode(this, node, nodes));
        }
        return nodes.values();
    }

    public Collection<FlowNode> postDominatorTree() {
        Map<ControlFlow.Node, FlowNode> nodes = new LinkedHashMap<>();
        for (ControlFlow.Node node : flow.postDominatorTree()) {
            nodes.put(node, new FlowNode(this, node, nodes));
        }
        return nodes.values();
    }

    public FlowBlock findBlockByIndex(int index) {
        for (FlowBlock block : blocks) {
            if (block.startIndex() == index) {
                return block;
            }
        }
        return null;
    }
}

File: src/main/java/io/disassemble/javanalysis/flow/ControlFlowGraph.kt
package io.disassemble.javanalysis.flow

import javassist.bytecode.analysis.ControlFlow

/**
 * @author Tyler Sedlar
 * @since 5/20/2017
 *
 * @param flow The [ControlFlow] for this graph to represent
 * @param blocks The list of [ControlFlow.Block]s within this graph
 */
class ControlFlowGraph(val flow: ControlFlow, val blocks: Array<ControlFlow.Block>) {

    /**
     * Retrieves the dominator tree of this [ControlFlowGraph].
     */
    val dominatorTree: Array<ControlFlow.Node>
        get() = flow.dominatorTree()

    /**
     * Retrieves the post dominator tree of this [ControlFlowGraph].
     */
    val postDominatorTree: Array<ControlFlow.Node>
        get() = flow.postDominatorTree()

    /**
     * Gets the block within this graph matching the given index.
     *
     * @param index The index of the block to find.
     *
     * @return The [ControlFlow.Block] within this graph matching the given index.
     */
    fun findBlockByIndex(index: Int): ControlFlow.Block ? {
        return blocks.firstOrNull { it.startIndex == index }
    }
}
